### PR TITLE
Adding event TogglingRadio(Player, RadioItem, bool, bool)

### PR DESCRIPTION
**Pull request by Rysik5318**

### CODE REVIEW

File: Exiled.Events/EventArgs/Player/TogglingRadioEventArgs.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        /// <summary>
+        ///     Gets the <see cref="API.Features.Items.Radio" /> which is being used.
+        /// </summary>
+        public Radio Radio { get; }
+
+        /// <inheritdoc/>
+        public Item Item => Radio;
+
+        /// <summary>
+        ///     Gets or sets a value indicating whether new state of the radio.
```
Comment (VALERA771) [inline Exiled.Events/EventArgs/Player/TogglingRadioEventArgs.cs:53]: Rewrite this pls

File: Exiled.Events/Patches/Events/Player/TogglingRadio.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                            b = 0;
+                        }
+
+                        PlayerChangeRadioRangeEvent playerChangeRadioRangeEvent = new PlayerChangeRadioRangeEvent(__instance.Owner, __instance, (RadioMessages.RadioRangeLevel)b);
+                        if (!EventManager.ExecuteEvent(playerChangeRadioRangeEvent))
+                        {
+                            return false;
+                        }
+
+                        b = __instance._rangeId = (byte)playerChangeRadioRangeEvent.Range;
```
Comment (VALERA771) [inline Exiled.Events/Patches/Events/Player/TogglingRadio.cs:76]: I thnk `b` should be used somewhere after this

File: Exiled.Events/Patches/Events/Player/TogglingRadio.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        private static bool Prefix(RadioItem __instance, RadioMessages.RadioCommand command)
+        {
+            if (__instance.Owner.HasBlock(BlockedInteraction.ItemUsage))
+                return false;
+
+            switch (command)
+            {
+                case RadioMessages.RadioCommand.Enable:
+                    {
+                        if (__instance._enabled || __instance._battery <= 0f || !EventManager.ExecuteEvent(new PlayerRadioToggleEvent(__instance.Owner, __instance, true)))
```
Comment (VALERA771) [inline Exiled.Events/Patches/Events/Player/TogglingRadio.cs:36]: Should be of dev boi but I think our events should be executed before NW's

File: Exiled.Events/Patches/Events/Player/TogglingRadio.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                        if (!ev.IsAllowed)
+                            return false;
+
+                        __instance._enabled = ev.NewState == __instance._enabled ? true : ev.NewState;
+                        break;
+                    }
+
+                case RadioMessages.RadioCommand.Disable:
+                    {
+                        if (!__instance._enabled || !EventManager.ExecuteEvent(new PlayerRadioToggleEvent(__instance.Owner, __instance, false)))
```
Comment (VALERA771) [inline Exiled.Events/Patches/Events/Player/TogglingRadio.cs:50]: Same as above

**Review by VALERA771 — COMMENTED** [review]
Everything except this looks pretty good

File: Exiled.Events/Patches/Events/Player/TogglingRadio.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    /// <summary>
+    ///     Patches <see cref="RadioItem.ServerProcessCmd" />.
+    ///     Adds the <see cref="Handlers.Player.TogglingRadio" /> event.
+    /// </summary>
+    [EventPatch(typeof(Handlers.Player), nameof(Handlers.Player.TogglingRadio))]
+    [HarmonyPatch(typeof(RadioItem), nameof(RadioItem.ServerProcessCmd))]
+    internal static class TogglingRadio
+    {
+        private static bool Prefix(RadioItem __instance, RadioMessages.RadioCommand command)
```
Comment (louis1706) [inline Exiled.Events/Patches/Events/Player/TogglingRadio.cs:27]: i will see if @BoltonDev  can make you the Transpiler also yeah ExiledEvent need to be call before NWEvent if it's deniable

**Review by louis1706 — CHANGES_REQUESTED** [review]

**Comment by louis1706** [comment]
also always build test in Release and not Debug

**Comment by VALERA771** [comment]
Also iirc if you would use prefix like this, `ChangingRadioPreset` event won't call

File: Exiled.Events/Patches/Events/Player/TogglingRadio.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    [HarmonyPatch(typeof(RadioItem), nameof(RadioItem.ServerProcessCmd))]
+    internal static class TogglingRadio
+    {
+        private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
+        {
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
+
+            Label retLabel = generator.DefineLabel();
+
+            LocalBuilder ev = generator.DeclareLocal(typeof(TogglingRadioEventArgs));
```
Comment (BoltonDev) [inline Exiled.Events/Patches/Events/Player/TogglingRadio.cs:43]:
```suggestion
```
Can you remove this pls ? I forgot to remove it

**Review by louis1706 — APPROVED** [review]

File: Exiled.Events/EventArgs/Player/TogglingRadioEventArgs.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+namespace Exiled.Events.EventArgs.Player
+{
+    using API.Features;
+
+    using Exiled.API.Features.Items;
+    using Interfaces;
+    using InventorySystem.Items.Radio;
+
+    /// <summary>
+    ///     Contains all information before toggling a radio.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/TogglingRadioEventArgs.cs:17]:
```suggestion
    /// Contains all information before toggling a radio.
```

Remove all white spaces as they're useless and it's horrible.
Do that for all occurrences.

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

**Review by louis1706 — APPROVED** [review]

File: Exiled.Events/Patches/Events/Player/TogglingRadio.cs [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+                // this
+                new(OpCodes.Ldarg_0),
+
+                // true
+                new(OpCodes.Ldc_I4_1),
+
+                // true
+                new(OpCodes.Ldc_I4_1),
+
+                // TogglingRadioEventArgs ev = new TogglingRadioEventArgs(Player, Radio, bool);
```
Comment (BoltonDev) [inline Exiled.Events/Patches/Events/Player/TogglingRadio.cs:63]:
```suggestion
                // TogglingRadioEventArgs ev = new TogglingRadioEventArgs(Player, RadioItem, bool, bool);
```
Error on my part here and in the other below

Comment (BoltonDev) [inline Exiled.Events/Patches/Events/Player/TogglingRadio.cs:63]:
```suggestion
                // TogglingRadioEventArgs ev = new TogglingRadioEventArgs(Player, RadioItem, bool, bool);
```
Error on my part here and in the other below

**Pull request merged**